drivers/bt_ll_nrfxlib/flash/flash_ll_nrfxlib.c

```diff
@@ -19,13 +19,25 @@
 
 #define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
 
-/* NOTE: The driver supports unligned writes, but some file systems (like FCB)
+/* NOTE: The driver supports unaligned writes, but some file systems (like FCB)
  * may use the driver sub-optimally as a result. Word aligned writes are faster
  * and require less overhead. This value can be changed to 4 to minimize this
  * overhead.
  */
 #define FLASH_DRIVER_WRITE_BLOCK_SIZE 1
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, nordic_nrf51_flash_controller), okay)
+#define DT_DRV_COMPAT nordic_nrf51_flash_controller
+#elif DT_NODE_HAS_STATUS(DT_INST(0, nordic_nrf52_flash_controller), okay)
+#define DT_DRV_COMPAT nordic_nrf52_flash_controller
+#elif DT_NODE_HAS_STATUS(DT_INST(0, nordic_nrf53_flash_controller), okay)
+#define DT_DRV_COMPAT nordic_nrf53_flash_controller
+#elif DT_NODE_HAS_STATUS(DT_INST(0, nordic_nrf91_flash_controller), okay)
+#define DT_DRV_COMPAT nordic_nrf91_flash_controller
```

Comment:
... yes, this change to dts macros must have been definitely to improve the maintainability

Reply:
The DT_DRV_COMPAT construct certainly is strange.

Reply:
This style is used upstream because a maintainer there is allergic to node labels. You can just get a node identifier for the flash controller with DT_NODELABEL(flash_controller) on all our SoCs.

```
$ rg flash_controller: dts/arm/nordic
dts/arm/nordic/nrf52833.dtsi
61:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf52810.dtsi
44:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf9160_common.dtsi
7:flash_controller: flash-controller@39000 {

dts/arm/nordic/nrf51822.dtsi
48:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf52820.dtsi
50:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf52811.dtsi
48:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf52840.dtsi
60:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf5340_cpunet.dtsi
53:	flash_controller: flash-controller@41080000 {

dts/arm/nordic/nrf52832.dtsi
52:		flash_controller: flash-controller@4001e000 {

dts/arm/nordic/nrf5340_cpuapp_common.dtsi
7:flash_controller: flash-controller@39000 {
```

Then you can deal with the node directly instead, using the generic APIs, instead of INST APIs.

Reply:
Replaced with suggestion from offline chat.

Reply:
The fact that you cannot tell what to use by 5 seconds of reading existing samples says a lot about current state of DTS.

It reminds me of one logging system I have seen once in a userspace driver. Instead of using few macros that would enable/disable printing (this was linux so you could simply write to a file and vfs would handle everything regarding caching and stuff), there was a very complex system of log management with backends and threads to handle that. A lot of good SW engineering stuff with some complex algorithms. It probably took 2 months to implement. In the end probably 90% of this was never used, nobody really understood it or could maintain it after the author moved to other place. And it was a good source of bugs too.

Or another example. This time a build system created for a large project because apparently none of existing build systems was good enough. This was perfect, except for the fact that anybody joining the project had to learn how this thing worked. Until one day project development environment switched from Windows to Linux. Nothing worked and there was nobody to handle it as there were never resources available to maintain it and author moved to the other place.

Maybe there is a perfect reason why this DTS is as is but I have a feeling that it is more complicated then it should be.